mds: for an irreparable hard link scrub is not able to identify the damage

sajibreadd opened the pull request:
Fixes: https://tracker.ceph.com/issues/69345
Signed-off-by: Md Mahamudur Rahaman Sajib <mahamudur.sajib@croit.io>





<!--
  - Please give your pull request a title like

      [component]: [short description]

  - Please use this format for each git commit message:

      [component]: [short description]

      [A longer multiline description]

      Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
      Signed-off-by: [Your Name] <[your email]>

    For examples, use "git log".
-->

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

sajibreadd commented:
jenkins test make check

gregsfortytwo commented:
Can you please explain (in the commit message) how you're cleaning up these remote dentries in scrub? How can you validate the link validity if it's linked to an inode where the auth is a remote server?

sajibreadd commented:
> Can you please explain (in the commit message) how you're cleaning up these remote dentries in scrub? How can you validate the link validity if it's linked to an inode where the auth is a remote server?

Commit message is updated. Mostly followed the code already written in mds by understanding it from high level. Feel free to share any insights you found.

sajibreadd commented:
@gregsfortytwo Are you reviewing it?

sajibreadd commented:
> > > Can you please explain (in the commit message) how you're cleaning up these remote dentries in scrub? How can you validate the link validity if it's linked to an inode where the auth is a remote server?
> > 
> > 
> > @gregsfortytwo Commit message is updated. Mostly followed the code already written in mds by understanding it from high level(not much experience with cephFS). Feel free to share any insights you found.
> 
> So, the issue if I understand it is that once the primary inode is gone (corrupted, gone, etc..), scrubbing the secondary inode isn't detecting the damage, right? If that's the case, there are two options: a) scrub the tree where the primary inode resides or b) when scrub encounters a remote link, try to open the link which would detect the faulty inode and mark the inode as damaged. What this change implements, is option (b), right?

Yes, but if repair flag is on, it will just clean up the dentry. Otherwise, everytime user stat/or hit that inode it will create a damage message.

vshankar commented:
> > > > Can you please explain (in the commit message) how you're cleaning up these remote dentries in scrub? How can you validate the link validity if it's linked to an inode where the auth is a remote server?
> > > 
> > > 
> > > @gregsfortytwo Commit message is updated. Mostly followed the code already written in mds by understanding it from high level(not much experience with cephFS). Feel free to share any insights you found.
> > 
> > 
> > So, the issue if I understand it is that once the primary inode is gone (corrupted, gone, etc..), scrubbing the secondary inode isn't detecting the damage, right? If that's the case, there are two options: a) scrub the tree where the primary inode resides or b) when scrub encounters a remote link, try to open the link which would detect the faulty inode and mark the inode as damaged. What this change implements, is option (b), right?
> 
> Yes, but if repair flag is on, it will just clean up the dentry and cleanup damage(This part I am not absolutely confident about my code, I just followed the server's code of removing remote link). Otherwise, everytime user stat or hit that inode it will create a damage message.

[cc @mchangir who has worked on cephfs scrub lately]

From the commit message of this change:

> for a remote link, tried to open the dentry(this part of code is copied from the mds path_traversal)
which internally push the damage in the damage list if applicable. For damaged remote link, a callback
context is called if repair flag is on, which clean up the dentry. To clean up the dentry unlinked it
marked the dentry dirty and unlinked it from the mdcache(followed the code from Server::_link_remote_finish)

I'm not 100% what's being done here. From my understanding, scrub code can (try to) lookup the remote inode for any damage and add it to the damage table. To repair a damage, the scrub command should be forwarded to the inode auth mds.

sajibreadd commented:
> I'm not 100% what's being done here. From my understanding, scrub code can (try to) lookup the remote inode for any damage and add it to the damage table. To repair a damage, the scrub command should be forwarded to the inode auth mds.

As for this dentry, it is not pointing to any inode, more specifically `dnl->get_inode()` is returning null pointer, so how do I get the auth mds? That's why I marked the dentry dirty which eventually remove dentry(`*_head`) from the omap of the object associated with the parent directory. Can you also check the reproduction step described in the tracker?

sajibreadd commented:
> So, the question is, if scrub is run on a secondary hardlink directory with repair flag, what should be done? Should it repair the damaged (primary) inode which is in another directory? Scrub should at least identify the damage IMO.

@vshankar thanks for explaining. Even we scrub the primary inode, it will not recover (As from the test case I deleted the object, let me know if I am missing something). Then how should user get out of this loop. Because everytime user will stat that hardlink(which they assume it exists), it will create damage. Let's say we identify the damage, but what should be the way to get rid of this damage. That's why I was talking about removing the dentry from the dentry list of it's parent directory. (I need to study a bit)

sajibreadd commented:
> Right. So, its either via scrub (w/ repair flag), or if the damage ins't fixable by scrub, then recover the metadata using the repair tools.

So in your opinion, we shouldn't cleanup/repair it this way right? Then I will update the repair part and keep the identification part.

sajibreadd commented:
> Right. So, its either via scrub (w/ repair flag), or if the damage ins't fixable by scrub, then recover the metadata using the repair tools.

Also do you think `rm <hardlink>` should work here?(without using repair tool)

vshankar commented:
> > Right. So, its either via scrub (w/ repair flag), or if the damage ins't fixable by scrub, then recover the metadata using the repair tools.
> 
> Also do you think `rm <hardlink>` should work here?(without using repair tool)

That does a path-walk on the MDS and if the inode is missing, then are bets are off...

patrakov commented:
> > > Right. So, its either via scrub (w/ repair flag), or if the damage ins't fixable by scrub, then recover the metadata using the repair tools.
> > 
> > 
> > Also do you think `rm <hardlink>` should work here?(without using repair tool)
> 
> That does a path-walk on the MDS and if the inode is missing, then are bets are off...

Speaking on behalf of the affected user.

We know where the original inode was, and that directory has been deleted, presumably by a script. So we know that the hardlink is truly dangling.

I personally don't care about the exact way that we should follow to clear the damage. I.e., so that "cat log.run.1" says "No such file or directory", not "I/O error". It can be rm. It can be some other tool or procedure. But this procedure should exist.

If this helps, I would advise @sajibreadd to split this MR into two:

1. This MR should add detection of this kind of damage so that the scrub finds all instances of it. Right now, without the patch, it finds none. No attempts to repair should be added to the scrub code, as the scrub cannot know whether the original inode exists or not.
2. A separate MR (possibly to the documentation only) that outlines a path for cleaning out the known dangling hardlinks as if they never existed.

sajibreadd commented:
@vshankar Should it repair the damaged (primary) inode which is in another directory(if possible with repair flag)? Should we do that or keep just the identification part?

mchangir commented:
> @vshankar Should it repair the damaged (primary) inode which is in another directory(if possible with repair flag)? Should we do that or keep just the identification part in this PR?

In my opinion, scrub should **not** attempt to repair things from secondary access paths. If at all, only the path to the primary path should be dumped. But if the inode itself is missing then we won't have the backtrace to the original path either. So, Venky's suggestion to `rm <hardlink>` via appropriate tool should suffice to heal the damage.

sajibreadd commented:
> In my opinion, scrub should **not** attempt to repair things from secondary access paths. If at all, only the path to the primary path should be dumped. But if the inode itself is missing then we won't have the backtrace to the original path either. So, Venky's suggestion to `rm <hardlink>` via appropriate tool should suffice to heal the damage.

So, I will just keep the identification part that means after scrub user will get these broken links. 
According to this comment https://github.com/ceph/ceph/pull/61173#issuecomment-2592183843, recovery tool is not helping to remove that dangling link as the primary inode metadata object is deleted/lost. Then how can we clean up the damage.

vshankar commented:
> @vshankar Should it repair the damaged (primary) inode which is in another directory(if possible with repair flag)? Should we do that or keep just the identification part in this PR?

Let's keep this PR to identify the damage.

vshankar commented:
> > @vshankar Should it repair the damaged (primary) inode which is in another directory(if possible with repair flag)? Should we do that or keep just the identification part in this PR?
> 
> In my opinion, scrub should **not** attempt to repair things from secondary access paths. If at all, only the path to the primary path should be dumped. But if the inode itself is missing then we won't have the backtrace to the original path either. So, Venky's suggestion to `rm <hardlink>` via appropriate tool should suffice to heal the damage.

Hang on - there seems to be some confusion. I didn't suggest `rm` the harlink - with a damaged inode, that isn't possible via the mount. If we somehow allow that, `nlink` count in the inode is not longer valid. Note that, missing metadata can be recovered via scanning data pool objects (rebuilding the linkages using backtraces).

@mchangir I'm sure there is documentation on repairing metadata using scrub. Maybe it needs to be polished to include examples for various damages.

vshankar commented:
@mchangir @sajibreadd `scan_links` (using cephfs-data-scan) tool can be used to fixup incorrect `nlink` in inode. So, to workaround a dangling hardlink, remove its omap entry and run `scan_links`.

vshankar commented:
> > @mchangir @sajibreadd `scan_links` (using cephfs-data-scan) tool can be used to fixup incorrect `nlink` in inode. So, to workaround a dangling hardlink, remove its omap entry and run `scan_links`.
> 
> Does `scan_links` require taking the filesystem offline?

Yes.

sajibreadd commented:
I removed the repair part.

vshankar commented:
> @vshankar we don't have the inode object in the data-pool for that corresponding file. So should it still recover using `scan-links`? Is there any high level overview you can share how `scan-link` will recover this dangling hard link damage(where inode object is missing/lost in the data pool)?

I can, but this discussion is best done outside this PR (slack). Let's catch over slack and I will explain the scan_{extents,inodes} and scan_links procedure. I'll be available post lunch...

sajibreadd commented:
> I can, but this discussion is best done outside this PR (slack). Let's catch over slack and I will explain the scan_{extents,inodes} and scan_links procedure. I'll be available post lunch...

Sure, just note that, the reproducer I showed in the tracker and the case I am currently talking about is different. In the reproducer I removed the inode metadata object of the directory which contains the primary file, but primary inode object of that file still exists in the data pool. But we have a case in our production cluster where we searched for the primary inode object in our data pool, we couldn't find it. So primary inode object is missing from the data pool. So as described in the recovery tools that it recovers metadata object from the data pool seems invalid here. That's why I am a bit curious. I will ping you in the slack.

sajibreadd commented:
jenkins test make check

mchangir commented on src/mds/ScrubStack.cc:
```diff
@@ -471,6 +471,21 @@ void ScrubStack::scrub_dir_inode_final(CInode *in)
 }
 
+void ScrubStack::identify_remote_link_damage(CDentry *dn) {
+  std::string path;
+  dn->make_path_string(path, true);
```
I don't see path used anywhere.

sajibreadd replied:
@mchangir updated now.

mchangir commented:
jenkins test make check

mchangir commented:
jenkins test make check arm64

mchangir commented on src/mds/ScrubStack.cc:
```diff
@@ -471,6 +471,19 @@ void ScrubStack::scrub_dir_inode_final(CInode *in)
+    if (mdcache->mds->damage_table.is_remote_damaged(dnl->get_remote_ino())) {
+      dout(4) << "scrub: remote dentry points to damaged ino " << *dn << dendl;
+      return;
+    }
+    mdcache->open_remote_dentry(dn, true, nullptr);
```
What is the intention of opening the remote dentry here ?

sajibreadd replied:
This will automatically log the damaged remote dentry(if it is) in the damage list. I followed it from path_traverse function.

mchangir reviewed (approved):
LGTM

vshankar commented on src/mds/ScrubStack.cc:
```diff
@@ -471,6 +471,19 @@ void ScrubStack::scrub_dir_inode_final(CInode *in)
+      dout(4) << "scrub: remote dentry points to damaged ino " << *dn << dendl;
+      return;
+    }
+    mdcache->open_remote_dentry(dn, true, nullptr);
```
Suggest to use `MDCache::get_dentry_inode()` that does the open underneath.

sajibreadd replied:
But this function does not have the check `if (mdcache->mds->damage_table.is_remote_damaged(dnl->get_remote_ino())) ` whether it already exists in the damage list or not?. Also it has a parameter `mdr`.
![image](https://github.com/user-attachments/assets/8d002392-8829-4aeb-8647-af9dbff99d02)
I think it already do those checks whether it is primary or remote in side `ScrubStack::scrub_dirfrag` function, so calling `get_dentry_inode` won't that do these checks again redundantly?

vshankar replied:
I revisited this today after a long delay. Sorry for that!

Doesn't passing `nullptr1 to `open_remote_dentry()` results in a fault since `_open_remote_dentry_finish` does `fin->complete(...)`?

vshankar reviewed (commented):
Looks acceptable. Also suggest to add a test case (should to trivial to test) and a bit of documentation related to scrub.

Nice work @sajibreadd

sajibreadd commented:
> Looks acceptable. Also suggest to add a test case (should to trivial to test) and a bit of documentation related to scrub.

@vshankar Not sure what kind of test and documentation to write here. Is it possible to elaborate more and where to write the test and which document need to be updated?

vshankar commented:
> > Looks acceptable. Also suggest to add a test case (should to trivial to test) and a bit of documentation related to scrub.
> 
> @vshankar Not sure what kind of test and documentation to write here. Is it possible to elaborate more and where to write the test and which document need to be updated?

Sure. CephFS qa suite has functional tests implemented as python test script. E.,g.: `qa/tasks/cephfs/test_scrub_checks.py` for scrub related tests. The tests are pretty self explanatory and adding a new test is pretty straightforward. For this change, you would add a tests closely following your reproducer - there are helpers functions that allow you to delete a particular RADOS object from the metadata pool and then invoke scrub. Let me know if you need help in adding it.

As far as documentation in concerned, you would want to add a note in `doc/cephfs/scrub.rst` regarding this functionality.

sajibreadd commented:
@vshankar In general I have a question,
```
void MDCache::open_remote_dentry(CDentry *dn, bool projected, MDSContext *fin, bool want_xlocked)
{
  dout(10) << "open_remote_dentry " << *dn << dendl;
  CDentry::linkage_t *dnl = projected ? dn->get_projected_linkage() : dn->get_linkage();
  inodeno_t ino = dnl->get_remote_ino();
  int64_t pool = dnl->get_remote_d_type() == DT_DIR ? mds->get_metadata_pool() : -1;
  open_ino(ino, pool,
      new C_MDC_OpenRemoteDentry(this, dn, ino, fin, want_xlocked), true, want_xlocked); // backtrace
}
```
You can see there is a function `open_ino`,
As it is a file we are sending pool = -1 as parameter.
```
void MDCache::open_ino(inodeno_t ino, int64_t pool, MDSContext* fin,
		       bool want_replica, bool want_xlocked,
		       vector<inode_backpointer_t> *ancestors_hint,
		       mds_rank_t auth_hint)
{
  dout(10) << "open_ino " << ino << " pool " << pool << " want_replica "
	   << want_replica << dendl;

  auto it = opening_inodes.find(ino);
  if (it != opening_inodes.end()) {
    open_ino_info_t& info = it->second;
    if (want_replica) {
      info.want_replica = true;
      if (want_xlocked && !info.want_xlocked) {
	if (!info.ancestors.empty()) {
	  CInode *diri = get_inode(info.ancestors[0].dirino);
	  if (diri) {
	    frag_t fg = diri->pick_dirfrag(info.ancestors[0].dname);
	    CDir *dir = diri->get_dirfrag(fg);
	    if (dir && !dir->is_auth()) {
	      filepath path(info.ancestors[0].dname, 0);
	      discover_path(dir, CEPH_NOSNAP, path, NULL, true);
	    }
	  }
	}
	info.want_xlocked = true;
      }
    }
    info.waiters.push_back(fin);
  } else {
    open_ino_info_t& info = opening_inodes[ino];
    info.want_replica = want_replica;
    info.want_xlocked = want_xlocked;
    info.tid = ++open_ino_last_tid;
    info.pool = pool >= 0 ? pool : default_file_layout.pool_id; //------> why?
    info.waiters.push_back(fin);
    if (auth_hint != MDS_RANK_NONE)
      info.auth_hint = auth_hint;
    if (ancestors_hint) {
      info.ancestors = std::move(*ancestors_hint);
      info.fetch_backtrace = false;
      info.checking = mds->get_nodeid();
      _open_ino_traverse_dir(ino, info, 0);
    } else {
      do_open_ino(ino, info, 0);
    }
  }
}
```
if pool == -1 then why we are only searching in the default pool, why not other fs associated pools or the pool set as the xattr? For example, if for that specific file ceph.layout.pool attribute is explicitly set to a different pool, then the default pool(e.g. `cephfs_data`)  won't have the 0'th object, right?

vshankar commented:
> if pool == -1 then why we are only searching in the default pool, why not other fs associated pools or the pool set as the xattr? For example, if for that specific file ceph.layout.pool attribute is explicitly set to a different pool, then the default pool(e.g. `cephfs_data`) won't have the 0'th object, right?

The default data pool will _always_ have the 0th object which has the file backtrace and layout even if custom data pool layout is set.

vshankar commented:
@sajibreadd ping?

sajibreadd commented:
jenkins test api

sajibreadd commented:
@vshankar any thoughts?

sajibreadd commented:
@vshankar @mchangir separated the commit. Let me know if some adjustment required.

vshankar commented:
> @vshankar @mchangir Any update?

Sorry, no. I will have a look this week.

vshankar commented on qa/tasks/cephfs/test_scrub_checks.py:
```diff
@@ -548,3 +591,4 @@ def __init__(self, command, rc, json_out, errstring):
     def __str__(self):
         return "Admin socket: {command} failed with rc={rc} json output={json}, because '{es}'".format(
             command=self.command, rc=self.rc, json=self.json, es=self.errstring)
+
```
unrelated change. please revert this hunk.

vshankar commented on doc/cephfs/scrub.rst:
```diff
@@ -150,6 +150,8 @@ These above named MDS damages can be repaired by using the following command::
 If scrub is able to repair the damage, the corresponding entry is automatically
 removed from the damage table.
 
+Note: Scrub with repair flag can idetify the hard link damage but can not repair it.
```
nit: scrub invoked with `repair` option can identify an damaged hard link but not repair it.

vshankar commented on qa/tasks/cephfs/test_scrub_checks.py:
```diff
@@ -477,6 +477,48 @@ def test_scrub_split_dirfrags(self):
+        rados_obj_file = "{ino:x}.00000000".format(ino=file_ino)
+        rados_obj_dir = "{ino:x}.00000000".format(ino=dir_ino)
+        self.fs.flush()
+        time.sleep(20)
```
Why sleep?

vshankar commented on qa/tasks/cephfs/test_scrub_checks.py:
```diff
@@ -477,6 +477,48 @@ def test_scrub_split_dirfrags(self):
+        self.fs.rados(["rm", rados_obj_file], pool=self.fs.get_data_pool_name())
+        self.fs.rados(["rm", rados_obj_dir], pool=self.fs.get_metadata_pool_name())
+        self.fs.fail()
+        time.sleep(10)
```
Why sleep? I you need to verify that the MDSs are no longer active, then there are constructs in class Filesystem that have helpers for this purpose.

vshankar commented on qa/tasks/cephfs/test_scrub_checks.py:
```diff
@@ -477,6 +477,48 @@ def test_scrub_split_dirfrags(self):
+        test_dir_path = "test_dir"
+        self.mount_a.run_shell(["mkdir", test_dir_path])
+
+        file_dir_path = test_dir_path + "/file_dir"
```
use os.path.join()

vshankar commented on qa/tasks/cephfs/test_scrub_checks.py:
```diff
@@ -477,6 +477,48 @@ def test_scrub_split_dirfrags(self):
+
+        file_dir_path = test_dir_path + "/file_dir"
+        self.mount_a.run_shell(["mkdir", file_dir_path])
+        file_path = file_dir_path + "/test_file.txt"
```
ditto.

vshankar commented on qa/tasks/cephfs/test_scrub_checks.py:
```diff
@@ -477,6 +477,48 @@ def test_scrub_split_dirfrags(self):
+        self.fs.rados(["rm", rados_obj_dir], pool=self.fs.get_metadata_pool_name())
+        self.fs.fail()
+        time.sleep(10)
+        self.fs.set_joinable()
```
All this is doing is to startup the MDS with an empty cache. Why not failover the MDS in that case?

anthonyeleven commented on doc/cephfs/scrub.rst:
```diff
@@ -150,6 +150,8 @@ These above named MDS damages can be repaired by using the following command::
 If scrub is able to repair the damage, the corresponding entry is automatically
 removed from the damage table.
 
+Note: scrub invoked with ``repair`` option can identify an damaged hard link but not repair it.
```
Note: A scrub invoked with the

sajibreadd commented:
> Doesn't passing nullptr to open_remote_dentry() results in a fault since _open_remote_dentry_finish does fin->complete(...)?
@vshankar Initially I added a nullptr checks, but now I just replaced it with `C_MDSInternalNoop`.

anthonyeleven reviewed (approved):
Docs lgtm, I'm not qualified to approve the code.

sajibreadd commented:
@vshankar a friendly ping

vshankar reviewed (approved):
LGTM. Will run this through tests.

vshankar commented:
This PR is under test in [https://tracker.ceph.com/issues/73082](https://tracker.ceph.com/issues/73082).

vshankar commented:
jenkins test windows

vshankar commented:
This is nearly good to go 👍
